spec.md

```diff
@@ -49,6 +49,119 @@ The following will not be part of the specification:
+that have been considered for the development of this specification.
+
+In these scenarios, we keep the roles of event producer and event consumer
+distinct. A single application context can always take on multiple roles
+concurrently, including being both a producer and a consumer of events.
+
+1) Applications produce events for consumption by other parties, for instance
+   for providing consumers with insights about end-user activities, state
+   changes or environment observations, or for allowing complementing the
+   application's capabilities with event-driven extensions.
+
+   Events are typically produced related to a context or a producer-chosen
+   base classification. A temperature sensor in a room might be
+   context-qualified by mount position, room, floor, and building. A sports
+   result might be classified by league and team.
+
+   The producer application might reside on server hosts or on a device of
+   any kind.
+
+   The produced events might be rendered and emitted directly by the producer
+   or by an intermediary; as example for the latter, consider event data
+   transmitted by a device over payload-size-constrained networks such as
+   LoRaWAN or ModBus, and where events compliant to this
+   specification will be rendered by a network gateway on behalf of the
+   producer.
+
+2) Applications consume events originating from devices and/or people
+   and/or applications/-modules/-instances for display, archival, analytics,
+   workflow processing, or other handling.
+
+   Platform tools and applications consume events for the purpose of
+   monitoring the condition and/or providing transparency into the operation
+   of a business solution and its foundational building blocks.
+
+   The consumer application might reside on server hosts or on a device
+   of any kind.
+
+   A consuming application will typically be interested in:
+   - distinguishing events such that the exact same event is not
+     processed twice.
+   - identifying and selecting the origin context or the
+     producer-assigned classification.
+   - identifying the temporal order of the events relative to the
+     originating context and/or relative to a wall-clock.
+   - understanding the context-related detail information carried
+     in the event.
+
+   In some cases, the consuming application might be interested in:
+   - obtaining further details about the event's subject from the
+     originating context, like obtaining detail information about a
+     changed object that requires privileged access authorization.
+   - interact with the event's subject at the originating context,
+     for instance reading a storage blob after having been informed
+     that this blob has just been created.
+
+   Consumer interests motivate requirements for which information
+   producers ought to include an event.
+
+3) Middleware routes events from producers to consumers, or onwards
+   to other middleware. Applications producing events might delegate
```

Comment:
Suggested change: Middleware routes events from producers to consumers, or on to other middleware, without changing the event.

Reply:
> without changing the event.

Does that include the context attributes or just the payload?

Reply:
I think when middleware routes the event, it should include both the original context attributes and payload since the consumer has interest in the context attributes too. For example, some context attribute is needed to identify/extract a field in the payload.

Reply:
I was referring to _without changing_. Do we exclude that there might be context attributes that can be changed, removed or added by the middleware?

Reply:
"Without changing" would be contradicted by the details of the section, because transcoding and transforming is doubtlessly a change. The transform description has a clause on it that semantic changes are not allowed, and that includes omissions. 

On @deissnerk's point: I'd like to rename the "extensions" section we currently have in the spec to "annotations" and clarify that this section MAY be edited by intermediaries but MUST be forwarded. That would align with the notion that the AMQP TC arrived at, where the "bare message" which holds the content and core properties is immutable, but can be annotated in transit.

Reply:
> "Without changing" would be contradicted by the details of the section, because transcoding and transforming is doubtlessly a change. The transform description has a clause on it that semantic changes are not allowed, and that includes omissions.

+1 

> On @deissnerk's point: I'd like to rename the "extensions" section we currently have in the spec to "annotations" and clarify that this section MAY be edited by intermediaries but MUST be forwarded. That would align with the notion that the AMQP TC arrived at, where the "bare message" which holds the content and core properties is immutable, but can be annotated in transit.

@clemensv Some brief context: We chose `extensions` because it is term OpenAPI uses to describe custom attributes to its specification and we felt having some similarities between CloudEvents and OpenAPI would be helpful.  That said, I like your suggestion of having that section be editable.  It means that section will take on a new purpose, in addition to merely being a place for custom attributes.  I believe having a section containing editable attributes may solve a handful of problems.

Reply:
Is it true that middleware should *never* change the content of an event? A few examples I could think of:

- Scrubbing PII from an event before forwarding to downstream consumers;
- Enriching an event, such as hydrating an object on the event. Imagine a producer includes a `productId` in the event payload, then the middleware hydrates the `product` object based on that id to include product name, price, inventory, etc.

In both of these situations, I would still consider the original producer the `source` of the event rather than the middleware, as suggested by [this thread](https://github.com/cloudevents/spec/pull/117/files#r175473795).

Reply:
@alexdebrie middleware may modify (transcode/transform) the content/payload of the event as mentioned in the paragraphs below. The question is whether middleware can modify the common event metadata attributes?

Reply:
@lfourie Yea, I'm wondering if it's ever appropriate for middleware to modify the event payload before passing along. Example:

**Removing personally identifiable information (PII) from payload**

Perhaps my Users service emits a `user.created` event which includes the social security number of the newly created user. I want to configure my middleware to scrub that attribute before sending to downstream consumers.

From:

```json
{
  "firstName": "Alex",
  "lastName": "DeBrie",
  "dateOfBirth": "01-01-1980",
  "ssn": "555-55-5555",
}
```

to: 

```json
{
  "firstName": "Alex",
  "lastName": "DeBrie",
  "dateOfBirth": "01-01-1980",
}
```

In my mind, this is still a `user.created` event that's been emitted from my Users service rather than an entirely new event emitted by the middleware as suggested by @ultrasaurus's [object.created example](https://github.com/cloudevents/spec/pull/117#discussion_r175625483).

Reply:
@alexdebrie  IMO, if you're removing (or adding) anything, including PII, you're making a semantic change. A scrubbed event is no longer the original but rather originates from whoever does the scrubbing (or enriching); it is a derivative. A very basic test might be whether you can compute the same integrity hash over the payload values (not metadata or bits on wire) before and after transcoding or transform.

ESPECIALLY if you have PII involved, you'll want to have a clear break between the event data before and after the scrub, because you will want to have absolutely no doubt in any audit trails about what kind of data you're dealing with when the GDPR police shows up. 

All that does not at all mean that such an intermediary weren't allowed to publish the derived event as to relate to the *same context*. Your event doesn't originate from the "users" service, but it still relates to exact same user context. 

The event will have a different id (to delineate it from the unscrubbed original) and if you want to include provenance details in the event detail data, you might have a trail there that gives the id of the original event and a hash of that original event. 

This is actually brilliant example for why "source" isn't the right concept in the current spec, because your scenario calls for an original and a derivative event to be posted related to the exact same context by different actors for different segments of a system.